source/common/upstream/subset_lb.cc

```diff
@@ -0,0 +1,497 @@
+      subset_keys_(subsets.subsetKeys()), original_host_set_(host_set),
+      original_local_host_set_(local_host_set) {
+  ASSERT(subsets.isEnabled());
+
+  // Create filtered default subset (if necessary) and other subsets based on current hosts.
+  update(host_set.hosts(), {}, false);
+
+  // Configure future updates.
+  host_set.addMemberUpdateCb([this](const std::vector<HostSharedPtr>& hosts_added,
+                                    const std::vector<HostSharedPtr>& hosts_removed) {
+    update(hosts_added, hosts_removed, false);
+  });
+
+  if (local_host_set) {
+    // No need to invoke update with local = true: subset creation handles the local hosts.
+    local_host_set->addMemberUpdateCb([this](const std::vector<HostSharedPtr>& hosts_added,
+                                             const std::vector<HostSharedPtr>& hosts_removed) {
+      update(hosts_added, hosts_removed, true);
+    });
+  }
+}
+
+HostConstSharedPtr SubsetLoadBalancer::chooseHost(LoadBalancerContext* context) {
+  if (context) {
+    const Router::MetadataMatchCriteria* match_criteria = context->metadataMatchCriteria();
+    if (match_criteria) {
+      // Route has metadata match criteria defined, see if we have a matching subset.
+      SubsetPtr subset = findSubset(match_criteria->metadataMatchCriteria());
+      if (subset != nullptr && subset->lb_ != nullptr) {
+        HostConstSharedPtr host = subset->lb_->chooseHost(context);
+        // Fallback if the subset was empty (inactive), but not, for
+        // example, if all its hosts are unhealthy.
+        if (host || !subset->empty()) {
+          stats_.lb_subsets_selected_.inc();
+          return host;
+        }
+      }
+    }
+  }
+
+  if (default_host_subset_ == nullptr) {
+    return nullptr;
+  }
+
+  stats_.lb_subsets_fallback_.inc();
+  return default_host_subset_->lb_->chooseHost(context);
+}
+
+// Iterates over the given metadata match criteria (which must be
+// lexically sorted by key) and find a matching SubsetPtr, if any.
+SubsetLoadBalancer::SubsetPtr SubsetLoadBalancer::findSubset(
+    const std::vector<Router::MetadataMatchCriterionConstSharedPtr>& match_criteria) {
+  LbSubsetMap* subsets = &subsets_;
+  for (uint32_t i = 0; i < match_criteria.size(); i++) {
+    const Router::MetadataMatchCriterion* match_criterion = match_criteria[i].get();
+    const auto& subset_it = subsets->find(match_criterion->name());
+    if (subset_it == subsets->end()) {
+      // No subsets with this key (at this level in the hierarchy).
+      break;
+    }
+
+    const ValueSubsetMap& vs_map = subset_it->second;
+    const auto& vs_it = vs_map.find(match_criterion->value());
+    if (vs_it == vs_map.end()) {
+      // No subsets with this value.
+      break;
+    }
+
+    const LbSubsetEntryPtr& entry = vs_it->second;
+    if (entry->host_subset_ != nullptr && i + 1 == match_criteria.size()) {
+      return entry->host_subset_;
+    }
+
+    subsets = &entry->children_;
```

Comment:
I don't fully grok what's going on here yet, but if this is a trie, can't this become null and then we would crash if there are still criteria to match?

Reply:
We never insert a null LbSubsetEntryPtr into the ValueSubsetMap, so `entry` won't be null. The `children_` member may be empty, in which case `subsets` refers to an empty map and on the next iteration, `subsets->find()` will return an end iterator, causing the loop to exit.

I think I can change this a little to make it clearer.